sources/bloodhound/src/results.rs

```diff
@@ -0,0 +1,182 @@
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub version: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub url: Option<String>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+pub enum CheckStatus {
+    /// Successfully verified to be in the expected state.
+    PASS,
+    /// Found to not be in the expected state.
+    FAIL,
+    /// Unable to verify state, manual verification required.
+    SKIP,
+}
+
+impl fmt::Display for CheckStatus {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Mode {
+    Automatic,
+    Manual,
+}
+
+impl fmt::Display for Mode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+/// CheckerMetadata contains the metadata about individual checkers. This data
+/// is used by bloodhound to discover details about the available checks and
+/// make decisions about including the checks based on input like the compliance
+/// level to evaluate.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CheckerMetadata {
+    pub name: String,
+    pub id: String,
+    pub level: u8,
+    pub title: String,
+    pub mode: Mode,
+}
+
+impl fmt::Display for CheckerMetadata {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let output = serde_json::to_string(&self).unwrap_or_default();
+        write!(f, "{}", output)
+    }
+}
+
+/// CheckerResult contains the results of a performed check.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CheckerResult {
+    pub status: CheckStatus,
+    pub error: String,
+}
+
+impl fmt::Display for CheckerResult {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let output = serde_json::to_string(&self).unwrap_or_default();
+        write!(f, "{}", output)
+    }
+}
+
+/// The Checker trait defines the interface for a compliance check. Checkers are
+/// expected to be able to provide metadata about the check it performs, and be
+/// able to execute that check and provide the results of its findings.
+///
+/// The expectation for bloodhound checkers is:
+/// - Check is performed normally, return JSON output from execute and exit 0
+/// - Check has failed validation, return JSON output from execute and exit 0
+/// - Check could not be performed, return error text to stderr and exit 1
+pub trait Checker {
+    fn metadata(&self) -> CheckerMetadata;
+    fn execute(&self) -> CheckerResult;
+}
+
+/// Common checker type for reporting manual check results.
+pub struct ManualChecker {
+    pub name: String,
+    pub id: String,
+    pub title: String,
+    pub level: u8,
+}
+
+impl Checker for ManualChecker {
+    fn execute(&self) -> CheckerResult {
+        CheckerResult {
+            error: "Manual check, see benchmark for audit details.".to_string(),
```

Comment:
The `ManualChecker` message is confusing. Does `Manual` mean that the user has to manually verify ID in the benchmark doc? Or where are the audit details?

Reply:
`Manual` is a term used in the CIS benchmarks, so it should be well known to someone interested in benchmark compliance. It basically means "the compliance state of this item cannot be automatically determined and you will need to perform manual verification auditing to evaluate it yourself." The audit details are in the benchmark and we do not want to duplicate any of that here.